src/style-spec/function/definitions/index.js

```diff
@@ -0,0 +1,409 @@
+        static opName() { return 'has'; }
+        static type() { return BooleanType; }
+        static signatures() { return [[StringType, nargs(1, ObjectType)]]; }
+        compileFromArgs(args) {
+            return `this.has(${args.length > 1 ? args[1] : 'props'}, ${args[0]}, ${args.length > 1 ? 'undefined' : '"feature.properties"'})`;
+        }
+    },
+    'at': class extends CompoundExpression {
+        static opName() { return 'at'; }
+        static type() { return typename('T'); }
+        static signatures() { return [[NumberType, array(typename('T'))]]; }
+        compileFromArgs(args) { return fromContext('at', args); }
+    },
+    'length': class extends CompoundExpression {
+        static opName() { return 'length'; }
+        static type() { return NumberType; }
+        static signatures() {
+            return [
+                [array(typename('T'))],
+                [StringType]
+            ];
+        }
+        compileFromArgs(compiledArgs) {
+            const t = this.args[0].type;
+            assert(t.kind === 'array' || t.kind === 'primitive');
+            return `${compiledArgs[0]}.length`;
+        }
+    },
+
+    // // feature and map data
+    'properties': class extends CompoundExpression {
+        static opName() { return 'properties'; }
+        static type() { return ObjectType; }
+        static signatures() { return [[]]; }
+        compile() {
+            return 'this.as(props, "Object", "feature.properties")';
+        }
+    },
+    'geometry_type': class extends CompoundExpression {
+        static opName() { return 'geometry_type'; }
+        static type() { return StringType; }
+        static signatures() { return [[]]; }
+        compile() {
+            return 'this.get(this.get(feature, "geometry", "feature"), "type", "feature.geometry")';
+        }
+    },
+    'id': class extends CompoundExpression {
+        static opName() { return 'id'; }
+        static type() { return ValueType; }
+        static signatures() { return [[]]; }
+        compile() {
+            return 'this.get(feature, "id", "feature")';
+        }
+    },
+    'zoom': class extends CompoundExpression {
+        static opName() { return 'zoom'; }
+        static type() { return NumberType; }
+        static signatures() { return [[]]; }
+        static parse(args, context) {
+            const ancestors = context.ancestors.join(':');
+            // zoom expressions may only appear like:
+            // ['curve', interp, ['zoom'], ...]
+            // or ['let', ..., ['coalesce', ['curve', interp, ['zoom'], ...], ... ] ]
+            if (
+                !/\[2\]$/.test(context.key) ||
+                !/^(let\.result:|coalesce:)*curve$/.test(ancestors)
+            ) {
+                throw new ParsingError(
+                    context.key,
+                    'The "zoom" expression may only be used as the input to a top-level "curve" expression.'
+                );
+            }
```

Comment:
This check should happen in `function/index.js` rather than here, because this restriction on `[zoom]` only applies to expressions used for style property values (as opposed to, say, filters)